src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs

```diff
@@ -315,60 +382,49 @@ private static bool IsHostRelative(string path)
         {
-            if (ShouldConvertChallengeToForbidden())
+            // HandleForbidden by redirecting to AccessDeniedPath if set
+            if (Options.AccessDeniedPath.HasValue)
             {
-                // Handle 403 by redirecting to AccessDeniedPath if set
-                if (Options.AccessDeniedPath.HasValue)
+                try
                 {
-                    try
-                    {
-                        var accessDeniedUri =
-                            Request.Scheme +
-                            "://" +
-                            Request.Host +
-                            Request.PathBase +
-                            Options.AccessDeniedPath;
-
-                        var redirectContext = new CookieApplyRedirectContext(Context, Options, accessDeniedUri);
-                        Options.Notifications.ApplyRedirect(redirectContext);
-                    }
-                    catch (Exception exception)
-                    {
-                        var exceptionContext = new CookieExceptionContext(Context, Options,
-                            CookieExceptionContext.ExceptionLocation.ApplyResponseChallenge, exception, ticket: null);
-                        Options.Notifications.Exception(exceptionContext);
-                        if (exceptionContext.Rethrow)
-                        {
-                            throw;
-                        }
-                    }
+                    var accessDeniedUri =
+                        Request.Scheme +
+                        "://" +
+                        Request.Host +
+                        Request.PathBase +
+                        Options.AccessDeniedPath;
+
+                    var redirectContext = new CookieApplyRedirectContext(Context, Options, accessDeniedUri);
+                    Options.Notifications.ApplyRedirect(redirectContext);
                 }
-                else
+                catch (Exception exception)
                 {
-                    Response.StatusCode = 403;
+                    var exceptionContext = new CookieExceptionContext(Context, Options,
+                        CookieExceptionContext.ExceptionLocation.ApplyResponseChallenge, exception, ticket: null);
+                    Options.Notifications.Exception(exceptionContext);
+                    if (exceptionContext.Rethrow)
+                    {
+                        throw;
+                    }
                 }
-                return;
+                return Task.FromResult(true);
             }
-
-            if (Response.StatusCode != 401 || !Options.LoginPath.HasValue )
+            else
             {
-                return;
-            }
-
-            // Automatic middleware should redirect on 401 even if there wasn't an explicit challenge.
-            if (ChallengeContext == null && !Options.AutomaticAuthentication)
-            {
-                return;
+                return base.HandleForbiddenAsync(context);
```

Comment:
And if base.handleForbidden just set 403 and returned true, that wouldn't let any other handlers execute afterwards, which isn't really what we are trying to enforce here is it?